tools/boardgen.py

```diff
@@ -0,0 +1,556 @@
+                continue
+            yield pin
+
+    # Allows a port to add additional command-line arguments to be handled.
+    def extra_args(self, parser):
+        pass
+
+    # Load board->cpu mapping from csv.
+    def parse_board_csv(self, filename):
+        with open(filename, "r") as csvfile:
+            rows = csv.reader(csvfile)
+            for linenum, row in enumerate(rows):
+                try:
+                    # Skip empty lines, and lines starting with "#".
+                    if len(row) == 0 or row[0].startswith("#"):
+                        continue
+
+                    # Lines must be pairs of names.
+                    if len(row) != 2:
+                        raise PinGeneratorError("Expecting two entries in each row")
+                    board_pin_name, cpu_pin_name = (x.strip() for x in row)
+
+                    # All rows must include a cpu name.
+                    cpu_hidden = False
+                    if cpu_pin_name.startswith("-"):
+                        cpu_hidden = True
+                        cpu_pin_name = cpu_pin_name[1:]
+                    self._pin_type.validate_cpu_pin_name(cpu_pin_name)
+                    pin = self.find_pin_by_cpu_pin_name(cpu_pin_name, create=True)
+                    pin._available = True  # It's in pins.csv so must be available.
+                    pin._hidden = cpu_hidden  # Optionally don't make available to Python.
+
+                    # Rows can optionally alias to a board name.
+                    if board_pin_name:
+                        board_hidden = False
+                        if board_pin_name.startswith("-"):
+                            board_hidden = True
+                            board_pin_name = board_pin_name[1:]
+                        self._pin_type.validate_board_pin_name(board_pin_name)
+                        pin.add_board_pin_name(board_pin_name, board_hidden)
+
+                # Inject "file:line: " into the exception.
+                except PinGeneratorError as er:
+                    raise PinGeneratorError("{}:{}: {}".format(filename, linenum, er))
+
+    def parse_af_csv(self, filename, header_rows=1, pin_col=0, af_col=1):
+        headings = {}
+        with open(filename, "r") as csvfile:
+            rows = csv.reader(csvfile)
+            for linenum, row in enumerate(rows):
+                try:
+                    # Skip empty lines, and lines starting with "#".
+                    if len(row) == 0 or row[0].startswith("#"):
+                        continue
+
+                    # Consume `header_rows` non-blank/comment rows at the start.
+                    if header_rows:
+                        if not headings:
+                            # If this is the first header row then initialise
+                            # the headings dict.
+                            for af_idx, header in enumerate(row[af_col:]):
+                                headings[af_idx] = header.strip()
```

Comment:
re: stm and missing AF columns, some chips don't have/use certain AF domains, eg F413 jumps from AF12 to AF15. Similarly, with a "fake" AF16 added for ADC, chips like F405 then jump from AF13 to AF16.
This code based on enumerate will give incorrect numbers to columns after any skipped columns if they're not present in the CSV. 

It would be fine to enforce csv files to always include all domains, though to avoid weird errors it would also be beneficial here to check the column header actually includes the expected number, or parse the number from the header rather than enumerate?
enforce:
```
if str(af_idx) not in header:
    raise PinGeneratorError(
        "Out of sequence AF heading {}".format(header)
    )
```

parse:
```
                            for column, header in enumerate(row[af_col:]):
                                af_idx = int(header.strip().replace("AF", ""))
                                headings[column] = (af_idx, header.strip())
```
...
```
                    for col, af in enumerate(row[af_col:]):
                        af = af.strip()
                        if not af:
                            continue
                        af_idx, heading = headings.get(col, (col, ""))
                        pin.add_af(af_idx, heading, af)
```

Reply:
I've added some commits to format all the af files consistently (as well as padding them as discussed elsewhere), and make-pins.py now checks the headers.